src/functions/InModuleScope.ps1

```diff
@@ -75,26 +75,48 @@
     )
 
     $module = Get-ScriptModule -ModuleName $ModuleName -ErrorAction Stop
+    $sessionState = Set-SessionStateHint -PassThru -Hint "Module - $($module.Name)" -SessionState $module.SessionState
 
-    # TODO: could this simply be $PSCmdlet.SessionState? Because the original scope we are moving from
-    # is the scope in which this command is running, right?
-    # $originalState = $Pester.SessionState
-    # $originalScriptBlockScope = Get-ScriptBlockScope -ScriptBlock $ScriptBlock
-
-    # try {
-    # $sessionState = Set-SessionStateHint -PassThru -Hint "Module - $($module.Name)" -SessionState $module.SessionState
-    # $Pester.SessionState = $sessionState
-    # Set-ScriptBlockScope -ScriptBlock $ScriptBlock -SessionState $sessionState
-
-    # do {
-    # Write-ScriptBlockInvocationHint -Hint "InModuleScope" -ScriptBlock $ScriptBlock
-    & $module $ScriptBlock @Parameters @ArgumentList
-    # } until ($true)
-    # }
-    # finally {
-    # $Pester.SessionState = $originalState
-    # Set-ScriptBlockScope -ScriptBlock $ScriptBlock -SessionStateInternal $originalScriptBlockScope
-    # }
+    $wrapper = {
+        param (
+            [Parameter(Mandatory = $true)]
+            [scriptblock]
+            ${Script Block},
+
+            [hashtable]
+            $___Parameters___ = @{ },
+
+            [object[]]
+            $___ArgumentList___ = @(),
+
+            [System.Management.Automation.SessionState]
+            ${Session State},
+
+            ${Set Dynamic Parameter Variable}
+        )
+
+        # This script block is used to create variables for provided parameters that
+        # the real scriptblock can inherit. Makes defining a param-block optional.
+
+        & ${Set Dynamic Parameter Variable} -SessionState ${Session State} -Parameters $___Parameters___
+
+        # define this in the current scope to be used instead of $PSBoundParameter if needed
+        # $PesterBoundParameters = if ($null -ne $___Parameters___) { $___Parameters___ } else { @{} }
+        & ${Script Block} @___Parameters___ @___ArgumentList___
+    }
+
+    Set-ScriptBlockScope -ScriptBlock $ScriptBlock -SessionState $sessionState
+    Set-ScriptBlockScope -ScriptBlock $wrapper -SessionState $sessionState
+    $splat = @{
+        'Script Block'                   = $ScriptBlock
+        '___ArgumentList___'             = $ArgumentList
+        '___Parameters___'               = $Parameters
+        'Session State'                  = $sessionState
+        'Set Dynamic Parameter Variable' = $SafeCommands['Set-DynamicParameterVariable']
```

Comment:
I don't think you need to use this, it's trying to avoid setting common parameters, because it's working together with the proxy function generated from a real function. With InModuleScoe we won't be getting those unless user explicitly specifies them, and then they would be surprised if they did not have them set.

This is a pattern to copy: 

https://github.com/pester/Pester/blob/c28bdffec9818a5ff1c03f8c5fa5607e919d6900/src/functions/Mock.ps1#L1115-L1164

Context is a hashtable of variableName -> value, as you would have from -Parameters

$______isInMockParameterFilter you don't need to define. 

& $private:______mock_parameters.ScriptBlock @______arguments here also define the parameters as second splat.

Change the debug logging to go to Runtime instead of Mock.

Reply:
Good point about the common-parameters. Attempt to reuse some logic, but I agree it adds unnecessary overhead calling the function and extra check.

Thoughts on whether to include `$PesterBoundParameters`? I commented it out for now as the name wasn't logic in this case considering the parameters are bound through a single `-Parameters`. Could consider `$_` like `Add-DataToContext` to provide the whole dictionary. Or just stick with variables-only?

Reply:
let's keep it just the variables for now.

Reply:
https://github.com/pester/Pester/blob/c28bdffec9818a5ff1c03f8c5fa5607e919d6900/src/functions/Mock.ps1#L1132-L1135
What's going on here?

Reply:
This is using a hashtable to pass a single parameter to the script block. This is to make it easier to pass the parameters to the scriptblock, so you don't have to add new ones to two places (param block and "argument list" when calling the scriptblock).

But you cannot splat from member on an object, so the values are saved into a variable. And then splatted.

It's a weird mix of historical approaches to this where ____ and single variable is used to avoid polluting the local scope, and to avoid naming conflicts, but $private should be avoiding all of that.

Reply:
> You cannot splat from member on an object, so the values are saved into a variable. And then splatted.

Just figured that one out the hard way. Was about to silently delete the question, hehe 😄 

Style question: Do you prefer `& $wrapper $ScriptBlock $Parameters $ArugmentList $sessionState`, the single hashtable-parameter or the one below? private-scoped parameters makes splatting a bit ugly.
```powershell
$splat = @{
    'private:ScriptBlock'    = $ScriptBlock
    'private:Parameters'     = $Parameters
    'private:ArgumentList'   = $ArgumentList
    'private:SessionState'   = $sessionState
}
& $wrapper @splat
```

Reply:
single hashtable, then list of parameters then splatting the parameters.

Reply:
Interesting that splatting considers the modifiers to be part of the name. Looking at that the parameter itself is actually called private:ScriptBlock, never noticed that. Luckily we only need those for internal purposes. (otherwise it is possible to remove the variable from the local scope to get the same effect with public parameters.